docs/proposals/nodesets.md

```diff
@@ -0,0 +1,157 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/nodesets.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NodeSets in Kubernetes
+
+**Author**: Justin Santa Barbara
+
+**Status**: Proposal.
+
+This document presents the design of the Kubernetes NodeSet, describes use cases, and gives an overview of the code.
+
+## Motivation
+
+The official bring-up procedure for a kubernetes cluster is to use kube-up.  This is a pretty complicated bash script
+that is largely separate for each provider.  It offers limited functionality: you cannot change the number of nodes
+directly, nor change the instance type used.  Many people resort to alternative methods of installation,
+to varying degrees of success.
+
+NodeSets put node launching under the control of the master.  They allow the operator to create
+and destroy NodeSets dynamically, or to change the number of instances in a NodeSet.  They
+allow for easier upgrades by launching nodes with different OS/k8s versions.
+
+This simplifies the Kubernetes installation, because the primary requirement is simply to launch a master.  Then
```

Comment:
Yes.  I believe NodeSets would be optional (you can still create Nodes directly), at least at first.  So bare metal or non-enlightened clouds would create nodes as they do today.

If we wanted to mandate NodeSets in future, we could still do that, even if in the worst case we had to create one NodeSet per node.  I think the way it could work for bare-metal would be that you would specify IPs / CIDRs for the nodes, just as on a cloud you would specify the instance type.  That said, I don't think we should mandate NodeSets unless we discover a compelling reason to do so.